### PR TITLE
switch to concurrent quick sort

### PR DESCRIPTION
**Pull request by jfcg**
Hi,

[sorty](https://github.com/jfcg/sorty) is an in-place concurrent quick sort implementation and is faster with just 4 goroutines.

I created 6gb test file with:
```
dd if=/dev/urandom of=file.dat bs=6M count=1024
```
sorty sorts in 9.41s on my Ryzen 5 1600, where old version sorts in 10.48s.

Cheers..

### CODE REVIEW

**Comment by afiodorov** [comment]
wow that's amazing, I will time your contribution and merge if it's faster. Thanks a lot!

File: main.go [hunk 1]
```diff
@@ -133,26 +43,49 @@ func main() {
...
+		return
+	}
+
+	var str []string
 
+	// identify lines in data
+	p := 0
+	for i := 0; i < len(data); i++ {
+		if data[i] == '\n' {
+			str = append(str, sixb.BtS(data[p:i+1]))
```
Comment (afiodorov) [inline main.go:63]: are you not worried about allocations here? I found that 

```
	for i := 0; i < len(m); i++ {	
		if m[i] == newLine {	
			numLines++	
		}	
	}	

	lines := Lines{data: m, slices: make([]Slice, 0, numLines)
```

speeds up data loading A LOT

Comment (jfcg) [reply main.go:63]: append is smart enough to amortize allocation cost. this way you only pass through data once.

Comment (afiodorov) [reply main.go:63]: `str = append(str, sixb.BtS(data[p:i+1]))`

this grows the str array by x2 everytime we need more memory, making loading very slow and memory hungry. I will go back to preallocating.

Comment (afiodorov) [reply main.go:63]: I can't load my data with this `append` approach - malloc complains. This why I ended up adding capacity.

Comment (afiodorov) [reply main.go:63]: (on 102gb of ram, file size - 44gb..)

Comment (jfcg) [reply main.go:63]: hmm, maybe GOGC=30 will help

File: main.go [hunk 2]
```diff
@@ -133,26 +43,49 @@ func main() {
...
+			p = i + 1
+		}
+	}
+	if p < len(data) {
+		str = append(str, sixb.BtS(data[p:])) // last line
+	}
+
+	// in-place concurrent sorting, no extra memory, 4 goroutines
+	sorty.Mxg = 4
```
Comment (afiodorov) [inline main.go:72]: should this be `	sorty.Mxg = uint32(runtime.NumCPU())`?

File: main.go (hunk 1 above)
Comment (afiodorov) [reply main.go:63]: I'll just live with an extra pass through data for now instead of relying on GC config...

File: main.go (hunk 2 above)
Comment (jfcg) [reply main.go:72]: if you want to use all cores, yes.

File: main.go (hunk 1 above)
Comment (afiodorov) [reply main.go:63]: It's loaded now, the overhead of string slices is same as I had before, needs 16 bytes per line (I presume 8 for len and 8 for capacity? not sure)

Comment (jfcg) [reply main.go:63]: string header occupies 16 bytes, yes same size with your initial Slice type

Comment (afiodorov) [reply main.go:63]: thanks for confirming. How did you find this out? source code of go?

Comment (jfcg) [reply main.go:63]: unsafe.Sizeof("") is 16, you also find more info in reflect package.

Comment (afiodorov) [reply main.go:63]: ah must be 8 bytes for start of the data array and 8 bytes for length

**Comment by afiodorov** [comment]
Your submission is slower than the radix search on my file. I strongly suspect you don't want to be using quick sort on the entire line, but just the prefix. Do you reckon the sorty library allows me to do this? Perhaps using the lessswap function.

Edit: has been taking >26 minutes now, the original binary would be finished in 19...

**Comment by jfcg** [comment]
You just modified 4 to NumCPU and it took longer? that is very strange.
For random data sorting initial 19 bytes or the whole line should not differ, but maybe the nature of your big data file is not random.

**Comment by jfcg** [comment]
Could you try the original pull request and report the sorting time (writing to hard disk can differ a lot from trial to trial) ?
Btw sorty benefits from optimization flag -B significantly.

**Comment by afiodorov** [comment]
> You just modified 4 to NumCPU and it took longer? that is very strange.
> For random data sorting initial 19 bytes or the whole line should not differ, but maybe the nature of your big data file is not random.

yes, I am using 16 cores and has been 42 minutes now and it is still sorting that file, so I'd rather terminate this now

**Comment by afiodorov** [comment]
> Could you try the original pull request and report the sorting time (writing to hard disk can differ a lot from trial to trial) ?
> Btw sorty benefits from optimization flag -B significantly.

trying now with `go build --gcflags='-B'`, if doesn't complete within 25 minutes I terminate...

**Comment by jfcg** [comment]
great, can you observe all 16 cores busy in top ?

**Comment by afiodorov** [comment]
> great, can you observe all 16 cores busy in top ?

yep; btw the difference in timings between quicksort and radix implementations are definitely in sorting; by observing cpu load I know when it starts and complete... I will give you sort time for radix a bit later... need to rerun

**Comment by afiodorov** [comment]
>nature of your big data file is not random.

first 19 bytes are RFC3339 date...

and I don't need to sort by other fields, so sort should ideally ignore the rest of the line after first 19 bytes

**Comment by jfcg** [comment]
Interesting result I got
rmshot
10gb random file
rm1 is old version + sort timing
rm2 is the pull request with 4 goroutines.
the second is slightly faster in sorting but slow in writing the data to hd. interesting..
(also old version writes one extra new line at the output)

**Comment by jfcg** [comment]
ok I see now. radix sort returns a string list that is physically close in memory because it creates a second copy out of the first unsorted lines.
quicksort sorts the string headers, but the original file data stays intact. when you want to write the str list into the file, you are making many random accesses in the original file data. that is why the whole operation is slower :D

**Comment by afiodorov** [comment]
ok, I tried with `-B` flag. It's been 25 minutes and the file is still being sorted, my radix version would have already written the output. So I am terminating...

**Comment by jfcg** [comment]
so it is a trade-off, in-place sorting a list of strings will not touch the original data. if your memory is sufficient then radix sort seems the right choice here. if you want to sort a list close to your ram amount then in-place sort is probably mandatory..

**Comment by afiodorov** [comment]
>it creates a second copy out of the first unsorted lines.

my RAM usage remains ~63 GB with radix version; if it were a copy I'd need more ram no?

**Comment by jfcg** [comment]
it is a hybrid method [apparently](https://github.com/twotwotwo/sorts/blob/master/radixsort.go).
it also uses quicksort for smaller ranges. so maybe not a whole copy of the data then, i guess like 1.5 times the data.

**Comment by jfcg** [comment]
with 16 cores, what were the sort times for your data? (like in the screenshot "sorted in")

File: main.go (hunk 1 above)
Comment (afiodorov) [inline main.go:63]: I think this implementation where newline is included into the string can be problematic when the last line of the file doesn't have the new line and it is reshuffled, then 2 lines will be printed into one line

**Comment by afiodorov** [comment]
```
2020/05/10 08:36:27 Creating memory-mapped file...
2020/05/10 08:42:07 Created memory-mapped file in 5m40.735858598s. Splitting lines...
2020/05/10 08:44:03 Splitted file into lines in 1m55.439408643s. Sorting...
2020/05/10 08:52:42 Sorted in 8m38.98795428s. Writing...
2020/05/10 08:57:18 Wrote in 4m35.894711623s. Closing... # this is inaccurate now, ignores time takes to flush... moved timing elsewhere...
```

Timings from https://github.com/afiodorov/radixmmap/pull/2

File: main.go (hunk 1 above)
Comment (jfcg) [reply main.go:63]: Yes you are right, last line also needs newline, added.
This has the advantage of less number of I/O calls when writing the file. You are making an extra call to append newline.

Comment (afiodorov) [reply main.go:63]: Good point, including in https://github.com/afiodorov/radixmmap/pull/3

**Pull request closed without merging**